File: src/core/effective-config.ts
/**
 * Deterministic shallow overlay for a resolved Module effective configuration.
 *
 * Implements control-plane configuration §1.1 (`REQ-CFG-003`, `REQ-CFG-005`):
 * one deterministic, shallow overlay of the Extension-level object and the
 * Module object. Module top-level members replace or insert Extension members
 * in their entirety; there is no recursive object merge, arrays are replaced
 * (never concatenated), and JSON `null` is an ordinary replacement value, not a
 * delete instruction. Absent members inherit; present members always replace.
 *
 * Sources are never mutated. Both sources must be closed JSON objects (`JsonValue`
 * with object root); the resolved object is validated against the same profile
 * (no NaN, infinity, negative zero, `undefined`, functions, class instances,
 * cycles, or unpaired surrogates) before digesting. The digest is the JCS
 * (RFC 8785) SHA-256 of the resolved object via the existing canonical-json
 * layer, emitted as `sha256:<64 hex>`.
 */

import {
  assertJsonValue,
  canonicalJsonDigest,
  isJsonObject,
  type JsonValue,
} from "./canonical-json.js";

/** Maximum allowed top-level members of one resolved Module effective config. */
export const EFFECTIVE_CONFIG_MAX_PROPERTIES = 1024;

/** Failure codes for effective-config resolution. */
export type EffectiveConfigErrorCode =
  | "EFFECTIVE_CONFIG_SOURCE_INVALID"
  | "EFFECTIVE_CONFIG_MAX_PROPERTIES";

/** Error thrown when an effective config cannot be resolved per §1.1. */
export class EffectiveConfigError extends Error {
  readonly code: EffectiveConfigErrorCode;

  constructor(code: EffectiveConfigErrorCode, message: string) {
    super(message);
    this.name = "EffectiveConfigError";
    this.code = code;
  }
}

/**
 * The resolved Module effective object and its JCS digest. Effective child
 * values are shared by reference with the validated sources, which are never
 * mutated by this module; callers must treat the result as immutable to
 * preserve the digest contract.
 */
export interface EffectiveConfigOverlay {
  readonly effectiveConfig: Readonly<Record<string, JsonValue>>;
  readonly digest: string;
}

function assertSourceObject(
  value: unknown,
  path: string,
  label: string,
): asserts value is Readonly<Record<string, JsonValue>> {
  try {
    assertJsonValue(value, path);
  } catch (error) {
    if (error instanceof EffectiveConfigError) throw error;
    throw new EffectiveConfigError(
      "EFFECTIVE_CONFIG_SOURCE_INVALID",
      `${label} configuration is not closed JSON: ${(error as Error).message}`,
    );
  }
  if (!isJsonObject(value)) {
    throw new EffectiveConfigError(
      "EFFECTIVE_CONFIG_SOURCE_INVALID",
      `${label} configuration must be a JSON object`,
    );
  }
}

/**
 * Compute the shallow overlay `effective_config(m)` and its JCS digest.
 *
 * Overlay is a pure function of the two sources: extension members first, then
 * Module members replace or insert. Any resolved object above
 * {@link EFFECTIVE_CONFIG_MAX_PROPERTIES} members is rejected before return,
 * mirroring the candidate rejection that must precede revision commit.
 */
export function normalizeEffectiveConfig(
  extensionConfig: unknown,
  moduleConfig: unknown,
): EffectiveConfigOverlay {
  assertSourceObject(extensionConfig, "$.extension_config", "Extension");
  assertSourceObject(moduleConfig, "$.module_config", "Module");

  const overlay: Record<string, JsonValue> = {};
  for (const key of Object.keys(extensionConfig)) {
    overlay[key] = extensionConfig[key]!;
  }
  for (const key of Object.keys(moduleConfig)) {
    overlay[key] = moduleConfig[key]!;
  }

  const memberCount = Object.keys(overlay).length;
  if (memberCount > EFFECTIVE_CONFIG_MAX_PROPERTIES) {
    throw new EffectiveConfigError(
      "EFFECTIVE_CONFIG_MAX_PROPERTIES",
      `resolved effective_config has ${memberCount} top-level members, exceeding the ${EFFECTIVE_CONFIG_MAX_PROPERTIES} limit`,
    );
  }

  return { effectiveConfig: overlay, digest: canonicalJsonDigest(overlay) };
}

File: crates/dolly-core-reducer/src/effective_config.rs
//! Pure effective-config normalizer and member-ceiling gate.
//!
//! Establishes the one-way authority direction for Module configuration
//! (control-plane spec, `docs/spec/control-plane/01` §1.1, `REQ-CFG-003`,
//! `REQ-CFG-004`, `REQ-CFG-005`): the frozen Extension-level object is the
//! basis, the Module-level object is the exact user override source, and
//! `effective_config(m)` is their deterministic **shallow** overlay — every
//! top-level member of `e.config`, then for `m.config` each member either
//! replaces or inserts the same-named member in its entirety. There is no
//! recursive object merge, no array concatenation, and JSON `null` is an
//! ordinary replacement value rather than a delete instruction. The result is
//! emitted in JCS form and hashed; the inputs are never mutated. Both source
//! objects and the resolved object share the 1,024-member ceiling; an
//! over-limit overlay is rejected before any consumer (reducer `InstallConfig`
//! admission) can commit a configuration revision.

use dolly_canonical_json::canonicalize;
use serde_json::Map;
use serde_json::Value;

/// Maximum number of top-level members in a source object or in the resolved
/// effective Module configuration (`REQ-CFG-005` ceiling).
pub const MAX_EFFECTIVE_CONFIG_PROPERTIES: usize = 1_024;

/// Spec rejection reason for an over-ceiling effective configuration, emitted
/// as `ConfigurationCandidateRejected` details and as the consumer error code
/// (the vector literal is `effective_config_max_properties`; the consumer code
/// is the same term in the crate's uppercase error-code convention).
pub const EFFECTIVE_CONFIG_MAX_PROPERTIES: &str = "effective_config_max_properties";
pub const EFFECTIVE_CONFIG_MAX_PROPERTIES_CODE: &str = "EFFECTIVE_CONFIG_MAX_PROPERTIES";

/// Closed normalizer failure set. Every variant is a deterministic,
/// host-independent rejection; no fallback and no environment behavior.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum EffectiveConfigError {
    /// The Extension-level config source is not a JSON object.
    ExtensionConfigNotObject,
    /// The Module-level config source is not a JSON object.
    ModuleConfigNotObject,
    /// The Extension-level source exceeds the 1,024-member input ceiling.
    ExtensionConfigMaxProperties,
    /// The Module-level source exceeds the 1,024-member input ceiling.
    ModuleConfigMaxProperties,
    /// The resolved overlay exceeds the 1,024-member consumer ceiling.
    EffectiveConfigMaxProperties,
    /// The overlay cannot be emitted as JCS (semantic nesting too deep or a
    /// value outside the strict JSON profile).
    EffectiveConfigNotCanonicalizable,
}

/// Owned, canonical effective configuration plus its JCS digest — the durable
/// source an InstallConfig admission binds (`effective_config` +
/// `effective_config_digest`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NormalizedEffectiveConfig {
    pub effective_config: Value,
    pub effective_config_digest: String,
}

/// Computes the effective config for one Module instance as `e.config` overlaid
/// by `m.config` using top-level replace-not-merge semantics, then emits the
/// result in JCS form and hashes it.
///
/// Fail-closed when either source is not a JSON object or when any of the two
/// sources/resolved object exceeds the ceiling. Non-commutative by
/// construction: `m.config` overlay applies after `e.config`; there is no
/// reverse merge, cross-extension, alias, or recursive union anywhere in the
/// API.
pub fn normalize_effective_config(
    extension_config: &Value,
    module_config: &Value,
) -> Result<NormalizedEffectiveConfig, EffectiveConfigError> {
    let extension = extension_config
        .as_object()
        .ok_or(EffectiveConfigError::ExtensionConfigNotObject)?;
    let module = module_config
        .as_object()
        .ok_or(EffectiveConfigError::ModuleConfigNotObject)?;
    if extension.len() > MAX_EFFECTIVE_CONFIG_PROPERTIES {
        return Err(EffectiveConfigError::ExtensionConfigMaxProperties);
    }
    if module.len() > MAX_EFFECTIVE_CONFIG_PROPERTIES {
        return Err(EffectiveConfigError::ModuleConfigMaxProperties);
    }
    // Shallow overlay: the Extension basis, then replace/insert each Module
    // member in its entirety. `Map` is BTree-backed, so order is deterministic.
    let mut merged: Map<String, Value> = extension.clone();
    for (key, value) in module {
        merged.insert(key.clone(), value.clone());
    }
    if merged.len() > MAX_EFFECTIVE_CONFIG_PROPERTIES {
        return Err(EffectiveConfigError::EffectiveConfigMaxProperties);
    }
    let effective_config = Value::Object(merged);
    let (_, digest) = canonicalize(&effective_config).map_err(|_| {
        // A source that cannot be emitted in JCS form cannot be admitted as
        // configuration; fail closed rather than guess.
        EffectiveConfigError::EffectiveConfigNotCanonicalizable
    })?;
    Ok(NormalizedEffectiveConfig {
        effective_config_digest: digest.to_canonical_string(),
        effective_config,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn ceiling_inputs_rejected_independently() {
        let huge = {
            let mut map = Map::new();
            for index in 0..(MAX_EFFECTIVE_CONFIG_PROPERTIES + 1) {
                map.insert(format!("key-{index:04}"), json!(index));
            }
            Value::Object(map)
        };
        let small = json!({"a": 1});
        assert_eq!(
            normalize_effective_config(&huge, &small),
            Err(EffectiveConfigError::ExtensionConfigMaxProperties)
        );
        assert_eq!(
            normalize_effective_config(&small, &huge),
            Err(EffectiveConfigError::ModuleConfigMaxProperties)
        );
    }

    #[test]
    fn non_object_inputs_fail_closed() {
        assert_eq!(
            normalize_effective_config(&json!([1]), &json!({"a": 1})),
            Err(EffectiveConfigError::ExtensionConfigNotObject)
        );
        assert_eq!(
            normalize_effective_config(&json!({"a": 1}), &json!("x")),
            Err(EffectiveConfigError::ModuleConfigNotObject)
        );
    }

    #[test]
    fn overlay_is_replace_not_recursive_or_concatenative() {
        let extension = json!({"nested": {"a": 1}, "list": [1], "score": 0});
        let module = json!({"nested": {"b": 2}, "list": [2], "score": 1});
        let resolved = normalize_effective_config(&extension, &module).unwrap();
        assert_eq!(
            resolved.effective_config,
            json!({"nested": {"b": 2}, "list": [2], "score": 1})
        );
        // Original sources are untouched.
        assert_eq!(
            extension,
            json!({"nested": {"a": 1}, "list": [1], "score": 0})
        );
        assert_eq!(module, json!({"nested": {"b": 2}, "list": [2], "score": 1}));
    }

    #[test]
    fn stateless_across_calls() {
        let extension = json!({"a": 1});
        let module = json!({"b": 2});
        let first = normalize_effective_config(&extension, &module).unwrap();
        let second = normalize_effective_config(&extension, &module).unwrap();
        assert_eq!(first, second);
    }
}
